### PR TITLE
feat: initial implementation of HomotopyContinuation interface

### PR DESCRIPTION
**Pull request by AayushSabharwal**
```julia
julia> @variables x y z
julia> eqs = [
           0 ~ x^2 + y^2 + 2x*y
           0 ~ x^2 + 4x + 4
           0 ~ y * z + 4x^2
       ]
julia> @mtkbuild sys = NonlinearSystem(eqs)
julia> prob = MTK.HomotopyContinuationProblem(sys, [x => 1.0, y => 1.0, z => 1.0], [])
julia> sol = HomotopyContinuation.solve(prob.hcsys; threading = false)
```

## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API

### CODE REVIEW

File: src/systems/nonlinear/nonlinearsystem.jl [hunk 1]
```diff
@@ -598,3 +598,25 @@ function Base.:(==)(sys1::NonlinearSystem, sys2::NonlinearSystem)
...
 end
+
+struct HomotopyContinuationProblem{H, U}
+    sys::NonlinearSystem
+    hcsys::H
+    u0::U
+end
+
+function HomotopyContinuationProblem(args...; kwargs...)
+    error("Requires HomotopyContinuationExt")
```
Comment (ChrisRackauckas) [inline src/systems/nonlinear/nonlinearsystem.jl:609]: Expand this 😅

Comment (AayushSabharwal) [reply src/systems/nonlinear/nonlinearsystem.jl:609]: Yeah there's probably a decent bit more work to be done here. I just opened the PR for visibility and as an RFC on implementation.

File: ext/MTKHomotopyContinuationExt.jl [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    mtkhsys = MTKHomotopySystem(nlfn.f, p, nlfn.jac, hvars, length(eqs))
+
+    prob = ModelingToolkit.HomotopyContinuationProblem{typeof(mtkhsys), typeof(u0)}(
+        sys, mtkhsys, u0)
+end
+
+function CommonSolve.solve(prob::ModelingToolkit.HomotopyContinuationProblem; kwargs...)
+    sol = HomotopyContinuation.solve(prob.hcsys; kwargs...)
+    rsols = HomotopyContinuation.real_solutions(sol)
```
[the remark this answers is not in the document]
Comment (AayushSabharwal) [reply ext/MTKHomotopyContinuationExt.jl]: The solution as returned from HomotopyContinuation is now part of the `original` field of `NonlinearSolution`. All the solutions of the polynomial system can be accessed from there.

**Comment by AayushSabharwal** [comment]
I think most of what's left now is documentation. It would also be nice to not require symbolic jacobians. I want to integrate that with `ADTypes` and `DifferentationInterface`, since they're "free" dependencies. ADTypes is pulled in via DI, and DI is pulled in via DiffEqCallbacks and NonlinearSolve, both of which are dependencies of MTK.

**Comment by ChrisRackauckas** [comment]
Leave any Jac change for a future PR. Contain the scope a bit here.

**Comment by isaacsas** [comment]
One thing that Catalyst does that might be nice is to also support rational functions by converting them into polynomial systems. I think @TorkelE set this up with @shashi at some point via a mix of stuff in Catalyst and Symbolics.

File: ext/MTKHomotopyContinuationExt.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (isaacsas) [reply ext/MTKHomotopyContinuationExt.jl]: Awesome. thanks!

**Comment by AayushSabharwal** [comment]
Could you point to what transformation you're referring to regarding rational functions?

**Comment by isaacsas** [comment]
You can always multiply through by the denominators to make the system polynomial. 

If you look at the Catalyst extension you can see what @TorkelE did. 

https://github.com/SciML/Catalyst.jl/blob/master/ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl

**Comment by AayushSabharwal** [comment]
Ah, I see what you mean now. Yeah that should be simple.

**Comment by isaacsas** [comment]
Note, there is no reason not to leave rational systems for a follow up. Just having it on polynomial systems to start is nice!

**Comment by TorkelE** [comment]
I haven't gone through properly, but one test case that def should be added is parametric exponents (e.g. x^n). If these are not handled properly (I.e.  replaced with their value before sent to HC) HC will fail.

**Comment by ChrisRackauckas** [comment]
What's left for this to be ready?

**Comment by AayushSabharwal** [comment]
Adding the parametric units test @TorkelE suggested, and figuring out why CI is failing

**Review by ChrisRackauckas — APPROVED** [review]

**Pull request merged**